images/bridge2/assistant/assistant.go

```diff
@@ -0,0 +1,144 @@
+package assistant
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/ajbouh/substrate/images/bridge2/tracks"
+	"github.com/ajbouh/substrate/images/bridge2/transcribe"
+)
+
+var recordAssistant = tracks.EventRecorder[*AssistantEvent]("assistant")
+
+type AssistantEvent struct {
+	SourceEvent tracks.ID
+	Name        string
+	Prompt      string
+	Response    string
+	Error       string
+}
+
+type Agent struct {
+	Assistants map[string]string
+}
+
+func (a *Agent) HandleEvent(annot tracks.Event) {
+	if annot.Type != "transcription" {
+		return
+	}
+	in := annot.Data.(*transcribe.Transcription)
+
+	if len(in.Segments) == 0 || len(in.Segments[0].Words) == 0 {
+		log.Printf("assistant: transcription had no segments or words")
+		return
+	}
+	text := strings.TrimSpace(in.Text())
+	names := a.matchAssistants(text)
+	log.Printf("assistant: matched %v for: %s", names, text)
+	for _, name := range names {
+		go a.respond(annot, name, text)
+	}
+}
+
+func (a *Agent) matchAssistants(text string) []string {
+	var matched []string
+	text = strings.ToLower(text)
+	for name := range a.Assistants {
+		if strings.Contains(text, name) {
+			matched = append(matched, name)
+		}
+	}
+	return matched
+}
+
+func (a *Agent) respond(annot tracks.Event, name, prompt string) {
+	out := AssistantEvent{
+		SourceEvent: annot.ID,
+		Name:        name,
+		Prompt:      prompt,
+	}
+	// TODO get speaker name from transcription
+	log.Printf("assistant: about to call %s", name)
+	resp, err := a.call(name, "user", prompt)
+	if err != nil {
+		log.Printf("assistant: %s error: %v", name, err)
+		out.Error = "error calling assistant"
+	} else {
+		log.Printf("assistant: %s response: %s", name, resp)
+		out.Response = resp
+	}
+	recordAssistant(annot.Span(), &out)
+}
+
+func (a *Agent) call(name, speaker, prompt string) (string, error) {
+	endpoint := a.Assistants[name]
+
+	maxTokens := 4096
+	systemMessage := strings.ReplaceAll(`
+A chat between ASSISTANT (named {}) and a USER.
+
+{} is a conversational, vocal, artificial intelligence assistant.
+
+{}'s job is to converse with humans to help them accomplish goals.
+
+{} is able to help with a wide variety of tasks from answering questions to assisting the human with creative writing.
+
+Overall {} is a powerful system that can help humans with a wide range of tasks and provide valuable insights as well as taking actions for the human.
+`, "{}", name)
+
+	req := ChatCompletionRequest{
+		MaxTokens: maxTokens - len(systemMessage),
```

Comment:
can you abstract out the token calculation here, because "len" is technically wrong and we should only really fix it in one spot eventually